### PR TITLE
ledger: redeemer and data values are Data, remove Sealed

### PR DESCRIPTION
**Pull request by michaelpj**
This is phases 3 and 4 of https://github.com/input-output-hk/plutus/issues/1465.

There's quite a lot in here. On the plus side, I got rid of the "unlifting" stuff in the book and a lot of `Sealed` juggling.

I'd appreciate feedback on a one thing in particular.

# Wrapping validators

The real type of validators is now `Data -> Data -> PendingTx -> Bool`. However, we don't want people to *write* that, and ideally we don't even want to really expose the idea that the arguments are `Data`. To that end I've written `wrapValidator :: (IsData a, IsData b) => (a -> b -> PendingTx -> Bool) -> (Data -> Data -> PendingTx -> Bool)`.

Now, you can call this on your validator when you write it, but then you end up with a top-level signature with the ugly `Data` things in it. Plus, if you're doing the typed script instance stuff, we want to have a validator with the "real" types in it.

So it's nice to just call it when you actually compile the validator, something like `$$(compile [|| wrapValidator validator ||])`. This works fine, except:
- It's a bit clunky if you're also parameterizing the validator, you need to eta-expand (or use a fancy composition operator).
- Because we're in a splice, TH won't let us use typeclass instances defined in the same file!

The latter is quite annoying, since we're not "really" using them, and the error is bad. There is a workaround: `let wrap = wrapValidator @DataType @RedeemerType in $$(compile [|| wrap validator ||])`. This is pretty ugly, but I don't have a better solution, so it's what I'm doing at the moment.

(The typed script instance stuff therefore also has to carry around a wrapper with it!)

You might think: can't we just do this generically in the library? No: we need to see the concrete typeclass instances, which means this needs to be done in the user's source code. We could maybe write some TH to do it.

On the plus side, we could also move the bit that turns validators returning `Bool` into validators returning `()` or throwing into the wrapping process, which would be good.

# Known stuff I will follow up on

Will make issues for these:
- Generating `IsData` instances.
- Move `check` into `wrapValidator`.
- Now validators all have the same type, we can require that type when converting to `ValidatorScript`, which will catch people who fail to wrap their validators.
- Rename `DataScript` and `RedeemerScript` since they aren't scripts any more.
    - What should we call "data scripts" now?
- Teach the state machine library to not rely on you knowing the state.

### CODE REVIEW

**Comment by j-mueller** [comment]
Hm, is it not possible do something similar to `Ledger.Typed.Tx` for wrapping validators? 

Something like `data Location = OnChain | OffChain` and a type family `Validator (l :: Location) a b` where `instance Validator OnChain a b = Data -> Data -> PendingTx -> Bool`. Although that doesn't give you any more than `wrapValidator` does already... so I don't know!

> What should we call "data scripts" now?

Maybe just call them state and input values, or state and input data

**Comment by michaelpj** [comment]
> Although that doesn't give you any more than wrapValidator does already... so I don't know!

Yeah. To have another go at stating the constraints:
- I want people to write their validators with the "deserialized" types, so that we can e.g. assert type coherence between their validators and redeemers etc.
- We can't generically compile `toData` - we need a concrete type. So this has to happen in user code.

I could shove this into `ScriptType`, I guess, i.e.
```
class ScriptType a where
    ...
    redeemerToData :: CompiledCode (RedeemerType a -> Data)
    dataToData :: CompiledCode (DataType a -> Data)
```
which is a bit like having `IsData` superclass constraints except "precompiled". I guess that's slightly nicer than the current version for things using `ScriptType` at least.

> Maybe just call them state and input values, or state and input data

I think "redeemer" is pretty established, we should probably stick with it. It also works without "script", so that's fine.

"Data value" sounds okay. I don't like "state value" because they're not *always* state.

**Comment by michaelpj** [comment]
> I could shove this into ScriptType, I guess

This doesn't actually work: this stuff has to be passed at the last possible moment, e.g. we can't do this abstractly for `StateMachine`. Aggravating.

**Comment by michaelpj** [comment]
Okay, here's my best suggestion: instead of defining a specialised `wrap` function outside the quote, we just get people to put the whole expression for the validator in there, e.g.

`$$(compile [|| validator ||]) where validator = wrapValidator mkValidator`

This avoids the need for the type applications at least, which is nice. Plus we can give a generic piece of help: "If GHC complains about the stage restriction, just lift the whole thing out."

File: plutus-playground-server/usecases/CrowdFunding.hs [hunk 1]
```diff
@@ -81,6 +82,15 @@ refundRange cmp =
...
     deriving (Generic, ToJSON, FromJSON, ToSchema)
 
+instance PlutusTx.IsData CampaignAction where
+    toData Collect = PlutusTx.Constr 0 []
+    toData Refund = PlutusTx.Constr 1 []
+    {-# INLINABLE fromData #-}
+    fromData (PlutusTx.Constr i [])
+        | i == 0 = Just Collect
+        | i == 1 = Just Refund
+    fromData _ = Nothing
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-playground-server/usecases/CrowdFunding.hs]: Yes, I considered the indexing formulation especially for things that are just enums. I'll probably add an `Enum`-like class at some point so we can do something like `deriving IsData via Enumerated`.

`toData`/`fromData` for sure form a prism. But I'm not sure that this helps us much: I don't want to try and import even a subset of `lens`, and I don't want our users to have to think about lenses/prisms. Would make them much easier to write? I haven't  thought of a way yet.

[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-playground-server/usecases/CrowdFunding.hs]: Yeah, I only think that having prisms might be a good idea if they simplify definitions, but I haven't been able to come up with something.

**Comment by effectfully** [comment]
> However, we don't want people to write that, and ideally we don't even want to really expose the idea that the arguments are `Data`.

I'm not sure I agree with this. It's a very explicit way of aligning things and being explicit is good. I'd probably just tell the users about `Data` rather than rely on TH or explain that instance resolution doesn't work in certain cases in a splice. Running some code inside `compile` and some code outside of it feels cryptic to me.

**Comment by kwxm** [comment]
> I'd probably just tell the users about `Data` rather than rely on TH or explain that instance resolution doesn't work in certain cases in a splice.

I tend to agree with this.  In many cases (he said optimistically) the data will be quite simple and it might be more straightforward if users could just write it explicitly rather than having some opaque translation applied in the background.  If they do that, then they might lose Haskell's type-safety guarantees and get the members of a pair the wrong way round or something, but then that would be their own fault.  Giving them the choice of whether to use `Data` explicitly or implicitly doesn't seem _too_ bad. 

There might also be a question of interoperability.  If someone uses another language for writing off-chain code and it has to produce outputs with datascripts which are compatible with ones produced by Haskell code then they'll have to know how the Haskell version is encoded, and that might not be too obvious if the encoding's been done automatically by some typeclass behind the scenes.

File: plutus-playground-server/usecases/CrowdFunding.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-playground-server/usecases/CrowdFunding.hs]: It seems like there should be a way, but it doesn't seem to be well studied. There is [this thing](https://github.com/frasertweedale/hs-fresnel), but that's it afaict.

**Comment by michaelpj** [comment]
> Running some code inside compile and some code outside of it feels cryptic to me.

To be clear, it's not running outside, it's just that the definition needs to be there. You can also make it work by putting the instance in another file!

> I tend to agree with this. In many cases (he said optimistically) the data will be quite simple and it might be more straightforward if users could just write it explicitly rather than having some opaque translation applied in the background.

Well, you can look at what was here before this PR for an example of that. It's just lots of noise IMO. 

You can still do it that way after this PR if you like, I just don't think it should be encouraged.

And it's not really opaque: you can look at the definition of `wrapValidator`, it's a very trivial function.

> There might also be a question of interoperability. If someone uses another language for writing off-chain code and it has to produce outputs with datascripts which are compatible with ones produced by Haskell code then they'll have to know how the Haskell version is encoded, and that might not be too obvious if the encoding's been done automatically by some typeclass behind the scenes.

Yes, as with all such serialization things, if you do this you probably want to hand-write your instances, but that's fine I think.

---

I think the more important thing is that in the long run we want people to be using the more type interfaces, and those can't work without having a function that takes the "deserialized" types so it can check that they're right.

**Comment by michaelpj** [comment]
To expand a little: people are going to *know* about `Data` (otherwise these `IsData` instances won't make much sense), I just don't expect them to deal with it in their validators.

**Comment by effectfully** [comment]
> I tend to agree with this. In many cases (he said optimistically) the data will be quite simple and it might be more straightforward if users could just write it explicitly rather than having some opaque translation applied in the background.

Sorry, I didn't mean that people should manipulate `Data` directly, that would quite cumbersome I think. Just that asking to apply `wrapValidator` directly and provide the type signature that mentions `Data` seems better than the other options to me.

> To be clear, it's not running outside, it's just that the definition needs to be there.

Sure, I should have said "writing" rather than "running".

> You can also make it work by putting the instance in another file!

Yes, but this is the kind of thing that I'd prefer to avoid explaining.

**Comment by michaelpj** [comment]
> Just that asking to apply wrapValidator directly and provide the type signature that mentions Data seems better than the other options to me.

To be clear, the alternative I'm envisaging is something like:

```
validator :: Data -> Data -> PendingTx -> Bool
validator = wrapValidator $ \d r p -> ...
```
vs
```
validator :: DataType -> RedeemerType -> PendingTx -> Bool
validator d r p =
```

The downsides of 1 as compared to 2 are:
- Where do the types of the inner validator come from? Either rely more on type inference, or ugly annotating of lambda arguments, or a new top-level binding (in which case we're back with 2).
- No typed validator binding to use in typed interfaces.

I guess I just really want people to live/think in the typed world as much as possible, and leave the "wrapping/parsing" business to the point where they package up their nice function for the library to use.

------

Anyway, I think I'm going to merge this as is in a bit. We can change that particular bit of the library fairly easily, and there's a bunch of uncontroversial stuff I want to do on top of this. Let's discuss this more at a meeting.

**Comment by effectfully** [comment]
> or a new top-level binding (in which case we're back with 2).

That's what I propose. Require the user to write two top-level bindings. Yes, this is slightly repetitive, but it's a small constant overhead per contract and zero magic while other solutions (generate the "untyped" validator via TH or apply it outside of a splice) are either magical or weird.

**Comment by michaelpj** [comment]
> That's what I propose. Require the user to write two top-level bindings. Yes, this is slightly repetitive, but it's a small constant overhead per contract and zero magic while other solutions (generate the "untyped" validator via TH or apply it outside of a splice) are either magical or weird.

Well, that's what this PR does. Except they're local bindings, not top-level bindings. No magic, really. I don't think making it a top-level binding significantly changes anything...

**Pull request merged**